### PR TITLE
fix(encryption): change message signing prefix

### LINKED ISSUE #1328: Incorrect prefix for message signing
We use the wrong phrasing for the message signing prefix:

https://github.com/hirosystems/stacks.js/blob/5ac9385069ce48bde452520d5168c84e31b5b961/packages/encryption/src/messageSignature.ts#L5-L7

Following the [Ethereum format](https://eips.ethereum.org/EIPS/eip-712#signatures-and-hashing-overview), already followed on the Ledger app, the message should be _Stacks Signed Message_, not _Stacks Message Signing_.

This is a change of length, too, so if I'm not mistaken about how the prefixing works, should be:

```
\x17Stacks Signed Message:\n
```

`"Stacks Signed Message:\n".length.toString(16) // 17`

Useful info on prefix here https://blog.ricmoo.com/verifying-messages-in-solidity-50a94f82b2ca

### PR DESCRIPTION
**Pull request by kyranjamie**
> This PR was published to npm with the version `4.3.9-pr.bb49829.0`
> e.g. `npm install @stacks/common@4.3.9-pr.bb49829.0 --save-exact`

Closes #1328 

This is a potentially breaking change for message signing, but brings the stacks implementation of message signing inline with the eth precedent, and stacks ledger app implementation

cc/ @aulneau for microstacks change?

### CODE REVIEW

**Review by beguene — APPROVED** [review]
Good catch ! Code looks good, prefix calculation looks good too.

**Comment by aulneau** [comment]
@kyranjamie i'm curious about this change, it would diverge from [SIP-018](https://github.com/stacksgov/sips/pull/57) have you brought this up there yet?

Also, what happens with any apps using this in production?

Can you say more about the rational of why this is important?

Edit: I now realize that there is nothing in the SIP associated with this lol -- but I do worry that there might be some breakage of apps with this change, what do you think?

**Comment by kyranjamie** [comment]
I think it's important we're consistent and follow existing patterns, unless there's a good reason not to. (What's the rationale for deviating?) 

Using the noun, _Signed Message_ means the prefix describes the content, rather than the verb "Message Signing", which is the action required to get the result.

There may be some breakages, but I'm thinking it's better to adjust this inline with our Ledger app now while the features relatively new, than leaving as is.

cc/ @friedger @pradel for input

**Review by fbwoolf — APPROVED** [review]

**Comment by kyranjamie** [comment]
The Latest ledger app, [`0.22.9`](https://github.com/Zondax/ledger-blockstack/releases/tag/v0.22.9), requires all payloads to begin with this prefix, rejecting those that use anything else.

**Review by janniks — APPROVED** [review]

**Comment by kyranjamie** [comment]
@janniks can you help get this one merged? Do you reckon this should be a `BREAKING CHANGE`?

**Comment by janniks** [comment]
How urgent is the timeline for this?
I would love to bundle it with some other breaking changes.

**Comment by kyranjamie** [comment]
> How urgent is the timeline for this? 

Not super urgent, waiting on Ledger release for this to be absolutely necessary

**Comment by janniks** [comment]
Any timeline on the ledger release?

**Comment by kyranjamie** [comment]
They said "around two weeks" on the 29th August

**Comment by janniks** [comment]
~When ready let's squash merge this into `next` with a `BREAKING CHANGE` commit body~

**Comment by janniks** [comment]
Should we think about somehow coordinating libs and wallets to launch this at the same time? The problem is that backend/frontends will need to match the `prefix` which the wallet is using to correctly verify signatures.
I believe, this will break login for: Gamma, Sigle, Console

**Comment by aulneau** [comment]
I was thinking so, it might also be of use to keep the legacy one around for a short time so both can be used as apps migrate away from it

**Comment by janniks** [comment]
The only thing I can think of is letting the `verifyMessageSignature` run twice in the background and return `true` if any of the two `prefixes` can be verified. Will also only work if the `message` is passed-in directly, rather than already encoded+hashed.
_Feels a bit incorrect, but safe-enough for potentially stopping a breakage in live sign-in systems_.

**EDIT**: prototype https://github.com/hirosystems/stacks.js/pull/1356

**Comment by kyranjamie** [comment]
Should we create a new method entirely `backwardsCompatibleVerifyMessageSignature` and release it with a lead on the wallet changes being released? Logging a warning message in the `verifyMessageSignature`?

**Comment by janniks** [comment]
Adding a backwards(/future) compatible function sounds interesting and more intentional (can be communicated to major sign-in projects). prototype: https://github.com/hirosystems/stacks.js/pull/1356/files

**Comment by aulneau** [comment]
I have implemented an update to the functions in `micro-stacks` as long as folks are on the latest, nothing should break when this feature lands

**Comment by ImTheCodeFarmer** [comment]
Is upgrading the packages enough to be up to date with this, or will any other changes be required?

**Comment by janniks** [comment]
**tl;dr** Yes, _more or less_  -- Basically, new messages will be signed with the new prefixes (also for the web-wallet, as soon as it upgrades). In addition, the verification method checks both the new and the old prefix.

---

For verification (`verifyMessageSignature`): the input should be a `signature`, a _trusted_ `publicKey`, a _plaintext/unhashed_ `message`. Then everything will work as expected. The method now compares against a hashed version of the `message` (twice / aka with both prefixes: old and new). If any of the two are valid, the method returns `true`.

In any case make sure to additionally check that the publicKey matches the address that is claims to be. [`publicKeyToAddress `](https://stacks.js.org/modules/transactions.html#publicKeyToAddress)

---

We'll also add more methods for these topics in the future to make them easier 🔐
